### PR TITLE
Add role checks (like IsScp) to Role Extensions / RoleTypeId

### PR DESCRIPTION
**Pull request by JayXTQ**
## Description
Add IsScp to Role Extensions to easily check if the Role is an SCP role, I noticed there was .IsHuman and .IsAlive, but no .IsScp which could be useful to some people

**What is the current behavior?** (You can also link to an open issue here)

You either have to do !RoleTypeId.IsHuman && RoleTypeId.IsAlive or RoleTypeId.Team == Team.SCPs to check if the team is SCPs

**What is the new behavior?** (if this is a feature change)

You can now check with RoleTypeId.IsScp to see if the team is an SCP team or not.

**Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

No, nobody will need to change anything if they don't want to.

**Other information**:

<br />

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentations
<br />

## Submission checklist

- [x] I have checked the project can be compiled
- [x] I have tested my changes and it worked as expected

### Patches (if there are any changes related to Harmony patches)
- [x] I have checked no IL patching errors in the console

### Other
- [ ] Still requires more testing

### CODE REVIEW

**Comment by JayXTQ** [comment]
No clue why this added 2 others, also no clue why the version got changed what

**Comment by JayXTQ** [comment]
Ok, so after discussing with moddedmcplayer, they said I should add IsHuman even though it exists natively inside PlayerRoles.RoleTypeId, should I add loads of other teams too based on what Exiled.API.Features.Player has before we merge this?

**Comment by moddedmcplayer** [comment]
Could be very subjective, discussing with Exiled maintainers.

**Comment by JayXTQ** [comment]
Wait so you'll add some of the methods or me? I don't mind making them. Just tell me what you think may be useful and I'll start

**Review by louis1706 — CHANGES_REQUESTED** [review]
Remove the change that removes "," character

**Review by louis1706 — CHANGES_REQUESTED** [review]
I don't think this is needed "here"

**Comment by louis1706** [comment]
Would be better to have it where you have the IsNtf property's

**Comment by JayXTQ** [comment]
Hey @louis1706 those characters being removed do nothing in actuality, just my code editor told me they should be removed so I removed them. Code still works fine. I will go ahead and move it to the IsNtf property

**Comment by JayXTQ** [comment]
Just had a look, there is no IsNtf property in this file. Should I move it beside `IsFpcRole`?

**Comment by louis1706** [comment]
> Hey @louis1706 those characters being removed do nothing in actuality, just my code editor told me they should be removed so I removed them. Code still works fine. I will go ahead and move it to the IsNtf property

They actually needed to prevent warning that put error in our repo because we wanted them to be put preventing useless changed line when you just add a new value under it

**Comment by louis1706** [comment]
> Just had a look, there is no IsNtf property in this file. Should I move it beside `IsFpcRole`?

Could you add it in both

**Comment by JayXTQ** [comment]
> > Hey @louis1706 those characters being removed do nothing in actuality, just my code editor told me they should be removed so I removed them. Code still works fine. I will go ahead and move it to the IsNtf property
> 
> They actually needed to prevent warning that put error in our repo because we wanted them to be put preventing useless changed line when you just add a new value under it

Sorry but I don't understand any of that

**Comment by JayXTQ** [comment]
> > Just had a look, there is no IsNtf property in this file. Should I move it beside `IsFpcRole`?
> 
> Could you add it in both

Add both IsNtf and IsScp? Ok I'll try add IsChaos and stuff too

**Comment by louis1706** [comment]
> > > Just had a look, there is no IsNtf property in this file. Should I move it beside `IsFpcRole`?
> > 
> > Could you add it in both
> 
> Add both IsNtf and IsScp? Ok I'll try add IsChaos and stuff too

I mean both file but that not bad too

**Comment by JayXTQ** [comment]
@louis1706 have a look at the changes now

**Review by moddedmcplayer — APPROVED** [review]

**Comment by VALERA771** [comment]
Won't this create a problem with ambiguous reference? Like I think that if yes then some plugins will require recompiling

**Comment by JayXTQ** [comment]
Unfortunately, yes I think it will, but blame @moddedmcplayer for that as he was adamant that I had to add IsHuman. What would you recommend @VALERA771

**Comment by moddedmcplayer** [comment]
> Won't this create a problem with ambiguous reference? Like I think that if yes then some plugins will require recompiling

Correct if I'm wrong but like quite certain the compiler turns extensions methods into normal calls..?
And then when editing it only happens when referencing both `PlayerRoles` and `API.Extensions`.

**Comment by moddedmcplayer** [comment]
> somebody

image
I'm not for replace, I am for adding wrappers and/or fill gaps that don't make sense. In this case calling base game is just not worth it.

**Comment by JayXTQ** [comment]
Yeah if PlayerRoles exists it makes sense for it to only use PlayerRoles.RoleTypeId.IsHuman as shown by the above screenshot. I don't think this will cause any errors honestly. But just let me know what you want me to do and I'll do it. But I just want this PR merged at some point lol

File: EXILED/Exiled.API/Features/Player.cs
Comment (louis1706) [inline EXILED/Exiled.API/Features/Player.cs:702]: Revert this change it's bad in my opinion

Comment (JayXTQ) [reply EXILED/Exiled.API/Features/Player.cs:702]: How so?

Comment (JayXTQ) [reply EXILED/Exiled.API/Features/Player.cs:702]: Also I was told to do this again, didn't do it as part of my own will. @moddedmcplayer what do you say about this review?

Comment (moddedmcplayer) [reply EXILED/Exiled.API/Features/Player.cs:702]: > Revert this change it's bad in my opinion

What we had before was duplicate code, even if it looks slightly better.

Comment (louis1706) [reply EXILED/Exiled.API/Features/Player.cs:702]: > > Revert this change it's bad in my opinion
> 
> What we had before was duplicate code, even if it looks slightly better.

it's still better to keep this one at what it's was before

Comment (JayXTQ) [reply EXILED/Exiled.API/Features/Player.cs:702]: So what is it that we are doing then?

Comment (JayXTQ) [reply EXILED/Exiled.API/Features/Player.cs:702]: @louis1706 @moddedmcplayer  please respond to the above ^^
I don't see the point in removing this code as now it fits better, we are now using exiled code to get these values
so I dont think we should revert this change lol

File: EXILED/Exiled.API/Extensions/RoleExtensions.cs
Comment (Mikihero) [inline EXILED/Exiled.API/Extensions/RoleExtensions.cs:152]: aren't guards also part of `Team.FoundationForces`?

Comment (JayXTQ) [reply EXILED/Exiled.API/Extensions/RoleExtensions.cs:152]: I think so, but guards are foundation forces anyway

File: EXILED/Exiled.API/Extensions/RoleExtensions.cs
Comment (Mikihero) [inline EXILED/Exiled.API/Extensions/RoleExtensions.cs:166]: currently this includes Guards because of the other thing I mentioned, I think you should specify that this doesn't (or does if it's intended) include them

File: EXILED/Exiled.API/Features/Player.cs
Comment (Mikihero) [reply EXILED/Exiled.API/Features/Player.cs:702]: > it's still better to keep this one at what it's was before

why exactly?

File: EXILED/Exiled.API/Features/Player.cs
Comment (Mikihero) [inline EXILED/Exiled.API/Features/Player.cs:707]: Just call it on the RoleType itself; `Role.Type.IsHuman()`

**Review by Mikihero — CHANGES_REQUESTED** [review]
imo these could be changed as well

File: EXILED/Exiled.API/Features/Player.cs
Comment (JayXTQ) [reply EXILED/Exiled.API/Features/Player.cs:707]: Doesn't work as the base game has IsHuman()

File: EXILED/Exiled.API/Extensions/RoleExtensions.cs
Comment (JayXTQ) [reply EXILED/Exiled.API/Extensions/RoleExtensions.cs:166]: It should grab guards as they are military, but if you want the docs to be changed to mention guards are included in both foundation and military I can do that

Comment (Mikihero) [reply EXILED/Exiled.API/Extensions/RoleExtensions.cs:166]: personally I wouldn't count guards as military, that's why I'd like you to specifically mention if they're counted or not

File: EXILED/Exiled.API/Extensions/RoleExtensions.cs
Comment (Mikihero) [reply EXILED/Exiled.API/Extensions/RoleExtensions.cs:152]: but they aren't Nine Tailed Fox, please correct

Comment (JayXTQ) [reply EXILED/Exiled.API/Extensions/RoleExtensions.cs:152]: Ok, I will remove them from Ntf

File: EXILED/Exiled.API/Extensions/RoleExtensions.cs
Comment (JayXTQ) [reply EXILED/Exiled.API/Extensions/RoleExtensions.cs:166]: Removing from IsNtf

Comment (Mikihero) [reply EXILED/Exiled.API/Extensions/RoleExtensions.cs:166]: If you could specify in its docs that doesn't include guards that'd be great as well since some people might be confused

File: EXILED/Exiled.API/Features/Player.cs
Comment (louis1706) [reply EXILED/Exiled.API/Features/Player.cs:707]: > Doesn't work as the base game has IsHuman()

if it's already exist we shouldn't make an other same extension because it's would create conflict

Comment (Misfiy) [reply EXILED/Exiled.API/Features/Player.cs:707]: This ^

Comment (Misfiy) [reply EXILED/Exiled.API/Features/Player.cs:707]: Also shouldnt need to modify it at all checking Role is HumanRole works and should never be incorrect

File: EXILED/Exiled.API/Features/Player.cs
Comment (Misfiy) [inline EXILED/Exiled.API/Features/Player.cs:696]: Keep it how it was

File: EXILED/Exiled.API/Features/Player.cs
Comment (Misfiy) [inline EXILED/Exiled.API/Features/Player.cs:690]: Keep

**Review by Misfiy — CHANGES_REQUESTED** [review]

File: EXILED/Exiled.API/Features/Player.cs
Comment (JayXTQ) [reply EXILED/Exiled.API/Features/Player.cs:690]: Should I just revert that whole file then or is there something I should keep because I am getting mixed everything over here. I don't know what is wanted and what is not wanted. Please someone just give me a straight answer.

**Review by Misfiy — APPROVED** [review]
Looks fine imo

**Comment by JayXTQ** [comment]
@louis1706 @Mikihero can you all check out these changes and let me know what you think.

**Review by Mikihero — APPROVED** [review]
lgtm

File: EXILED/Exiled.API/Extensions/RoleExtensions.cs
Comment (louis1706) [inline EXILED/Exiled.API/Extensions/RoleExtensions.cs:159]: Just does 
Ismilitary(this RoleTypeId roletype) =>  !roletype.IsCivilian()

I know this goes against what miki said but that what basegame believe for armor consumption of stamina for example

**Review by louis1706 — CHANGES_REQUESTED** [review]

File: EXILED/Exiled.API/Extensions/RoleExtensions.cs
Comment (VALERA771) [reply EXILED/Exiled.API/Extensions/RoleExtensions.cs:159]: Idk can we count tutorial as civilian (iirc when I checked it last time `RoleTypeId.Tutorial.IsCivilian()` will return false

Comment (JayXTQ) [reply EXILED/Exiled.API/Extensions/RoleExtensions.cs:159]: Tutorial is not civilian, no. so it will count as military, this change shouldn't be done @louis1706 as what I've done is perfect, this is just adding more work and more calculations for no reason.

**Review by VALERA771 — APPROVED** [review]

File: EXILED/Exiled.API/Extensions/RoleExtensions.cs
Comment (Mikihero) [reply EXILED/Exiled.API/Extensions/RoleExtensions.cs:159]: technically guard is a military class, I've said otherwise before because it doesn't make sense to me but that's how they're classified on the wiki, this method should reflect that

Comment (JayXTQ) [reply EXILED/Exiled.API/Extensions/RoleExtensions.cs:159]: Way to late to say that now @Mikihero 😭

Comment (Mikihero) [reply EXILED/Exiled.API/Extensions/RoleExtensions.cs:159]: to make it work you can just add `|| roletype == RoleTypeId.FacilityGuard`, I don't think there really is a better solution since negating `IsCivilian` would return shit like tutorial etc.

Comment (JayXTQ) [reply EXILED/Exiled.API/Extensions/RoleExtensions.cs:159]: yeah, but this PR has been merged, if I could directly commit I would

**Pull request merged**
